lib/src/component_declaration/component_base.dart

```diff
@@ -444,6 +444,18 @@ abstract class UiStatefulComponent<TProps extends UiProps, TState extends UiStat
   // ----------------------------------------------------------------------
 }
 
+class WarnOnModify<K, V> extends MapView<K, V> {
+  WarnOnModify(Map componentData): super(componentData);
+
+  @override
+  operator []=(K key, V value) {
+    assert(
+      false,
+      'Mutating the state or props directly will soon be impossible without causing breakages.'
```

Comment:
```suggestion
      '''
      Direct mutation of component props / state values directly is unsupported - and often 
      causes unexpected behavior(s) within UiComponent / UiStatefulComponent implementations.
      
      In the soon-to-be-released UiComponentV2 / UiStatefulComponentV2 classes, 
      direct mutation of these values will result in a runtime error.
      
      The mutation of props should be done by passing in a new value from a 
      parent / wrapper component or data store.
      
      The mutation of state should be done by calling setState(newState()..someKey = someValue).
      '''
```

@greglittlefield-wf I would love your thoughts on the accuracy / wording of this message.

Reply:
Looks good, but I think we should keep the error message terse that consumers don't glaze over it, and as few lines as possible so that it doesn't blow up the logs.

I think we should also include more information on what was mutated so that the consumer knows where to fix it. might also be nice to have more directed versions for props/state:
```
props["$key"] was updated incorrectly. Never mutate this.props directly, as it can cause unexpected behavior; props must be updated only by passing in new values when rerendering this component.

This will throw in UiComponentV2 (to be released as part of the React 16 upgrade).
```
```
state["$key"] was updated incorrectly. Never this.state directly, as it can cause unexpected behavior; state must be updated only via setState.

This will throw in UiComponentV2 (to be released as part of the React 16 upgrade).
```